enhancements/cli-in-cluster-management.md

```diff
@@ -0,0 +1,166 @@
+upstream effort we can use to provide an in-cluster CLI manager while also providing a mechanism that the Kubernetes community as a whole can utilize.  That brings us to option 4.
+
+4.  The CLI in-cluster manager can provide CLIs as plugins through [krew](https://github.com/kubernetes-sigs/krew).  CLI images will serve the artifacts to enable disconnected downloads, and OpenShift CLIs will be available as 
+[krew plugins](https://github.com/kubernetes-sigs/krew-index/tree/master/plugins).
+    * Krew and Krew plugins are upstream projects that Kubernetes users are already familiar with
+    * From the [README](https://github.com/kubernetes-sigs/krew/#what-does-krew-do): `Krew is a tool that makes it easy to use kubectl plugins. Krew helps you discover plugins, install and manage them on your machine. It is similar to tools like apt, dnf or brew. Today, over 70 kubectl plugins are available on Krew.` It makes sense to offer OpenShift CLIs in the same way.
+    * With krew, artifacts will be packaged within an image.  Each cluster will have an index of CLI images.  Mirrored local registries will provide binaries to disconnected environments.  
+
+Option 4 is the option we as a team are leaning toward.  It requires that each OpenShift CLI publish artifacts, for example, [index of /pub/openshift-v4/clients](https://mirror.openshift.com/pub/openshift-v4/clients/)
+then create an image similar to this: [example cli-artifacts image](https://github.com/openshift/oc/blob/master/images/cli-artifacts/Dockerfile.rhel).  CLI artifacts are packaged in images.
+The CLI in-cluster manager will gather information from each CLI's CR about each CLI and its image.  A Krew index will be populated in every cluster to provide available CLIs and plugins for a cluster.
+
+### Non-Goals
+
+* CLI manager will not build or serve the binaries.  It will know where to find them.
+* This proposal is not concerned with _which_ binaries will be managed.  This proposal is meant to determine the mechanism only.  Consumers and publishers are clients of the mechanism. 
+* CLI manager will not create or update the ClusterCLI Custom Resources, those will be managed by individual CLIs and their operators. 
+
+### Requirements:
+
+1. CLI author requirements:
+    - provide CLI image containing binaries
+    - Custom Resource that describes the binaries (location within the image, metadata, description)
+2. An operator that will be responsible for:
+    - registering the new CLI CRD 
+    - reading metadata from CR from authors and extracting the binaries from the CLI images to local disk
+    - serving extracted binaries
+    - serving krew-compatible index
+3. oc changes:
+    - embed krew
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a user, I want a CLI in-cluster manager for various CLIs available for OpenShift and related services so that I can discover, install and list them
+The user will invoke the following commands:
+
+* `oc krew list` to view installed plugins
```

Comment:
Isn't this going to be a little bit confusing? Most of the CLIs that we are talking about, are no `oc` plugins, but standalone CLIs. Using krew command might create a wrong impression.

Reply:
finally getting back to this, sorry for the delay - `krew` provides the path to install to a user's home directory - we can modify where we install within the home dir to allow for the standalone binary rather than a plugin.  We'll use `krew` for its established list|search|install|etc.. capabilities, but we'll modify for our needs.

Reply:
Would it be possible for clients/users without `oc` installed to to get information from cluster about available CLIs?

Reply:
yes, users will still be able to download/view from the console.

Reply:
Using console would require user to do download it manually. 
Lets say we have an IDE that uses various CLIs to interact with the cluster.  Would that be possible to avoid user interaction, except might be asking for username/password, and download required binaries from the cluster?

Reply:
without `oc` _and_ without the console? To download from the cluster you'd have to connect to the cluster.  You can always download from wherever the binaries are published (like https://mirror.openshift.com/pub/openshift-v4/clients/)